### ISSUE TITLE
extending specs allowing to extend dependencies

### ISSUE THREAD

**Issue by tkoomzaaskz** [pateketrueke/json-schema-faker#18]
I've seen that the specs are defined through json files, containing test-definitions, e.g.

```json
      {
        "description": "should handle format (faker)",
        "schema": {
          "type": "object",
          "properties": {
            "name": {
              "type": "string",
              "faker": "name.findName"
            },
            "email": {
              "type": "string",
              "faker": "internet.email"
            }
          },
          "required": ["name", "email"]
        },
        "valid": true
      },
```

It is very self-descriptive, but it inherits flaws of JSON: it cannot include functions. And if I want to add a function to extend faker/chance (see #12), I have to do something else...

I think these are the lines that cause problems: https://github.com/pateketrueke/json-schema-faker/blob/master/spec/main-spec.coffee#L41-L45. According to #12, I should be able to run:

```js
var jsf = require('json-schema-faker');
var schema = ...
jsf.customize('faker', function(faker){
    faker.locale = "de"; // or whatever
    faker.customModule = {
        fn_1: function...
        fn_2: function...
        fn_n: function...
    };
    return faker;
});
var sample = jsf(schema);
```
 so there should be something like:

```
          sample = try
            jsfaker(schema, refs)
            jsfaker.customize(function) <- ADDED
          catch e
            error = e.message
            throw e unless test.throws
```

**Comment by pateketrueke** [comment]
This [format.json](https://github.com/pateketrueke/json-schema-faker/blob/master/spec/core/issues/format.json#L17) spec is using the `formats()` method this way: [formats.coffee](https://github.com/pateketrueke/json-schema-faker/blob/master/spec/formats.coffee)

I think you must add a your module like [this](https://github.com/pateketrueke/json-schema-faker/blob/master/spec/main-spec.coffee#L6):

```coffeescript
require('./customize').register(jsfaker)
```

And the only method you should implements is `register()` unless you want to `validate()` your formats too.

I think this is enough yet.

**Comment by tkoomzaaskz** [comment]
Well, I think these two things are different:
* `jsf.format` is extending `json-schema-faker`
* extending `faker` which is the content of #12.
I would not mix those things.

Please take a look at https://github.com/tkoomzaaskz/json-schema-faker/tree/extending-dependencies. [Latest build](https://travis-ci.org/tkoomzaaskz/json-schema-faker/builds/61484324) is correct, it fails only due to 3rd party lib problem. I added:

        "customize": {
            "generator": "faker",
            "module": "core/customize/faker-extend"
        }

to the definition and 3 lines of code in main-spec.coffee. This is exactly what would be done in consumer code. What do you think about that?

**Comment by pateketrueke** [comment]
I like the idea, but using a "customize" key is too specific, we can do it more generic?

```json
"require": "core/customize/faker-extend"
```

And then encapsulate the calls to `formats()` or `customize()`, etc.

This way we can refactor the `format.json` with its own dependencies too and we can remove [this ugly line of code](https://github.com/pateketrueke/json-schema-faker/blob/master/spec/main-spec.coffee#L6).

> I've spent my day doing this...

Sorry for that, I never thought about those features.

**Comment by tkoomzaaskz** [comment]
I agree that "customize" key is too specific, you're right. And I like the idea of encapsulating it with "require". Just need to make sure that both functions: formats and customize accept the same headers. But I don't really know how you wanna do that. Customize accepts 2 parameters now: generator name and extending function...

Don't be sorry, I just want to make it work ;)

**Comment by pateketrueke** [comment]
> But I don't really know how you wanna do that. Customize accepts 2 parameters now: generator name and extending function...

We can do this?

```coffeescript
if test.require
  wrapper = require('./' + test.require)
  wrapper.register(jsfaker)
```

Of course you should wrap your factory for `customize()` like [formats.coffee](https://github.com/pateketrueke/json-schema-faker/blob/master/spec/formats.coffee).

**Comment by tkoomzaaskz** [comment]
Ok, I've done it. That was really hard, but finally, I did it. Please take a look. I encapsulated `customize` with `.register`. Can we publish it in `0.1.8`? I need it urgently :)

**Comment by pateketrueke** [comment]
Great job @tkoomzaaskz :beers:

I've published `0.1.8`!
